library/std/src/io/mod.rs

```diff
@@ -357,17 +357,30 @@ where
 // of data to return. Simply tacking on an extra DEFAULT_BUF_SIZE space every
 // time is 4,500 times (!) slower than a default reservation size of 32 if the
 // reader has a very small amount of data to return.
-pub(crate) fn default_read_to_end<R: Read + ?Sized>(r: &mut R, buf: &mut Vec<u8>) -> Result<usize> {
+pub(crate) fn default_read_to_end<R: Read + ?Sized>(
+    r: &mut R,
+    buf: &mut Vec<u8>,
+    size_hint: Option<usize>,
+) -> Result<usize> {
     let start_len = buf.len();
     let start_cap = buf.capacity();
+    // Optionally limit the maximum bytes read on each iteration.
+    // This adds an arbitrary fiddle factor to allow for more data than we expect.
```

Comment:
Given the comment immediately above this function, I don't think we want to tack on this fiddle factor, or at least not one this large. If the file appears to be 8 bytes long, we shouldn't be adding 1024 bytes and rounding up to DEFAULT_BUF_SIZE.

Reply:
Fair enough but this does not affect allocation in any way. This only has an effect if we already have a buffer that's way larger than we need. It places an upper limit on the length of the slice passed to the `read_buf` function. I figured if we already have a large buffer, then we may as well use *some* of that space if we need it.

Reply:
I see. And the caller can always pass a smaller buffer if it wants to, and `read_to_end` doesn't actually trigger this. OK, fair enough.